src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs

```diff
@@ -193,29 +188,39 @@ public byte[][] GetByteCodes(Keccak[] requestedHashes, long byteLimit)
             if (stopped || startingHash != Keccak.Zero)
             {
                 byte[][]? proofs = GenerateRangeProof(tree, startingHash, nodes[^1].Path, storageRoot);
-                return (responseNodes.ToArray(), proofs);
+                return (responseNodes, proofs);
             }
             responseSize += innerResponseSize;
         }
-        return (responseNodes.ToArray(), null);
+        return (responseNodes, null);
     }
 
     private (Dictionary<byte[], byte[]>?, long, bool) GetNodesFromTrieVisitor(Keccak rootHash, Keccak startingHash, Keccak limitHash,
         long byteLimit, long hardByteLimit, bool isStorage = false)
     {
         PatriciaTree tree = new(_store, _logManager);
-
-        RangeQueryVisitor visitor = new(startingHash.Bytes, limitHash.Bytes, !isStorage, byteLimit, hardByteLimit, HardResponseNodeLimit);
-        VisitingOptions opt = new() { ExpectAccounts = false, KeepTrackOfAbsolutePath = true };
-        tree.Accept(visitor, rootHash, opt);
-        (Dictionary<byte[], byte[]>? requiredNodes, long responseSize) = visitor.GetNodesAndSize();
-
-        return (requiredNodes, responseSize, visitor._isStoppedDueToHardLimit);
+        RangeQueryVisitor visitor = null;
+        try
+        {
+            visitor = new(startingHash.Bytes, limitHash.Bytes, !isStorage, byteLimit, hardByteLimit, HardResponseNodeLimit);
+            VisitingOptions opt = new() { ExpectAccounts = false, KeepTrackOfAbsolutePath = true };
+            tree.Accept(visitor, rootHash, opt);
+            (Dictionary<byte[], byte[]>? requiredNodes, long responseSize) = visitor.GetNodesAndSize();
+            return (requiredNodes, responseSize, visitor._isStoppedDueToHardLimit);
+        }
+        catch (Exception)
+        {
+            visitor?.Dispose();
+            throw;
+        }
+        finally
+        {
+            visitor?.Dispose();
+        }
```

Comment:
using instead of dispose. btw If you have something in finally you don't need it in catch.